src/backend/plaf.rs

```diff
@@ -0,0 +1,507 @@
+use std::{collections::HashMap, hash::Hash, rc::Rc};
+
+use halo2_proofs::halo2curves::ff::PrimeField;
+
+use crate::{
+    ast::{query::Queriable, ForwardSignal, InternalSignal, StepType, Trace},
+    compiler::{cell_manager::Placement, step_selector::StepSelector, FixedGenContext},
+    ir::{
+        Circuit as cCircuit, Column as cColumn,
+        ColumnType::{Advice as cAdvice, Fixed as cFixed, Halo2Advice, Halo2Fixed},
+        PolyExpr as cPolyExpr,
+    },
+    wit_gen::{GenericTraceContext, TraceWitness},
+};
+
+use num_bigint::BigUint;
+use polyexen::{
+    expr::{get_field_p, Column as pColumn, ColumnKind, ColumnQuery, Expr as pExpr, PlonkVar},
+    plaf::{
+        ColumnFixed, ColumnWitness, Lookup as pLookup, Plaf, Poly as pPoly, Witness as pWitness,
+    },
+};
+
+#[allow(non_snake_case)]
+pub fn chiquito2Plaf<F: PrimeField<Repr = [u8; 32]>, TraceArgs: Clone, StepArgs: Clone>(
+    circuit: cCircuit<F, TraceArgs, StepArgs>,
+    k: u32,
+    debug: bool,
+) -> (Plaf, ChiquitoPlafWitGen<F, TraceArgs, StepArgs>) {
+    let mut chiquito_plaf = ChiquitoPlaf::new(circuit.clone(), debug);
+    let plaf = chiquito_plaf.get_plaf(k);
+    let empty_witness = plaf.gen_empty_witness();
+    let wit_gen = ChiquitoPlafWitGen {
+        trace: circuit.trace,
+        placement: circuit.placement,
+        selector: circuit.selector,
+        step_types: circuit.step_types,
+        plaf_witness: empty_witness,
+        c_column_id_to_p_column_index: chiquito_plaf.c_column_id_to_p_column_index,
+    };
+
+    (plaf, wit_gen)
+}
+
+#[derive(Clone, Debug)]
+pub struct ChiquitoPlaf<F: PrimeField, TraceArgs, StepArgs: Clone> {
+    debug: bool,
+    circuit: cCircuit<F, TraceArgs, StepArgs>,
+    // Chiquito column id doesn't start from zero.
+    // Plaf column index starts from 0 for each column type (advice, fixed, and instance).
+    // Therefore a mapping is needed to convert chiquito column id to plaf index.
+    c_column_id_to_p_column_index: HashMap<u32, usize>,
+    advice_index: usize, // This is Plaf's index of witness (advice) column.
```

Comment:
do we use this variable for something? it seems it is assigned once but never used

Reply:
Fixed